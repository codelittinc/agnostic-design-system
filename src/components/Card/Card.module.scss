@import '~@codelitt/ay-css-library/src/components/brand/fonts';
@import '~@codelitt/ay-css-library/src/components/brand/colors';
@import '~@codelitt/ay-css-library/src/components/brand/grid';

.card-container {
  background-color: $ay-white-color;
  border: 1px solid $ay-grey-16-color;
  border-radius: 0.25rem;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: space-between;
  transition: box-shadow;
  width: 100%;

  .image-container {
    align-items: center;
    background-color: $ay-grey-8-color;
    display: flex;
    height: 9.5rem;
    justify-content: center;
    margin: 0.5rem 0.5rem 1rem;

    .image {
      height: 7.5rem;
      width: 7.5rem;
    }
  }

  .title {
    align-items: center;
    display: flex;
    margin: 1rem;
  }

  .description {
    margin: 0 1rem;
  }

  .shortcut {
    align-items: center;
    border-bottom: 1px solid $ay-grey-16-color;
    display: flex;
    height: 2rem;
    margin: 1rem 1rem 0;
    padding-bottom: 1rem;

    button {
      width: 100%;
    }

    &.no-border {
      border-bottom: 0;
    }
  }

  .button-container {
    width: 100%;
  }

  &:hover {
    .image-container {
      &.general {
        background-color: $ay-london-blue-color;
      }

      &.tenant {
        background-color: $ay-frozen-grass-color;
      }

      &.agency {
        background-color: $ay-pepa-pig-red-color;
      }
    }
  }
}

.primary-button {
  font-size: 1.12rem;
  font-weight: 500;
}

.card-title {
  font-size: 1.75rem;
  line-height: 2rem;
}

.card-description {
  font-size: 1rem;
  line-height: 1.5rem;
}

@media screen and (min-width: $view-width-small-desktop-min) {
  .card-container {
    &:hover {
      box-shadow: 0 0.5rem 1.5rem rgba(1, 14, 40, 0.24);
    }
  }
}

:export {
  view-width-small-desktop-min: $view-width-small-desktop-min;
}
