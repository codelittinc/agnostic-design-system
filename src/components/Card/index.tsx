import React, { useLayoutEffect, useState } from 'react';
import classnames from 'classnames';

import style from './Card.module.scss';

import Text from '../Text';
import Button from '../Button';

type ButtonType = {
  label: string;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
};

type PropTypes = {
  description: string;
  image: string;
  primaryButton: ButtonType;
  shortcuts: Array<ButtonType>;
  title: string;
  type?: 'general' | 'tenant' | 'agency';
  wrapperClassName?: string;
};

const getPxFromRem = (value: string) =>
  parseInt(value.replace('rem', ''), 10) * 16;

const minDesktopScreenSize = getPxFromRem(
  style['view-width-small-desktop-min'],
);
const maxCharactersSmallScreen = 70;
const maxCharactersBigScreen = 150;

const Card: React.FC<PropTypes> = (props: PropTypes) => {
  const {
    description,
    image,
    primaryButton,
    shortcuts,
    title,
    type,
    wrapperClassName,
  } = props;
  const [width, setWidth] = useState(window.innerWidth);

  useLayoutEffect(() => {
    const updateWidth = () => setWidth(window.innerWidth);

    window.addEventListener('resize', updateWidth);
    return () => window.removeEventListener('resize', updateWidth);
  }, []);

  const renderShortcut = (
    shortcut: ButtonType,
    index: number,
    isLastItem: boolean,
  ) => {
    return (
      <div
        className={classnames(style.shortcut, {
          [`${style['no-border']}`]: isLastItem,
        })}
        key={`${index}-${shortcut.label}`}
      >
        <Button
          iconPosition="left"
          label={shortcut.label}
          type="ghost"
          onClick={shortcut.onClick}
          size="small"
        />
      </div>
    );
  };

  const cardWith =
    width < minDesktopScreenSize
      ? maxCharactersSmallScreen
      : maxCharactersBigScreen;

  return (
    <div className={classnames(style['card-container'], wrapperClassName)}>
      <div>
        <div className={classnames(style['image-container'], style[type!])}>
          <img src={image} className={style.image} alt="card-icon " />
        </div>

        <div className={style.title}>
          <Text
            type="title"
            value={title}
            wrapperClassName={style['card-title']}
          />
        </div>
        <div className={style.description}>
          <Text
            value={description}
            maxLength={cardWith}
            wrapperClassName={style['card-description']}
          />
        </div>
      </div>

      <div>
        {shortcuts.map((shortcut, index) =>
          renderShortcut(shortcut, index, index === shortcuts.length - 1),
        )}
        <div className={style['button-container']}>
          <Button
            type="inline"
            label={primaryButton.label}
            onClick={primaryButton.onClick}
            wrapperClassName={style['primary-button']}
          />
        </div>
      </div>
    </div>
  );
};

Card.defaultProps = {
  type: 'general',
};
export default Card;
