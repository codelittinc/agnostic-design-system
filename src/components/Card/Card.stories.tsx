import * as React from 'react';
import { storiesOf } from '@storybook/react';
import Card from './';
import { action } from '@storybook/addon-actions';
import { select, text, withKnobs } from '@storybook/addon-knobs';
import NewsIcon from '../../assets/images/news.png';
import { Container, Row, Col } from '../Grid';

const stories = storiesOf('Components|Card', module);

stories.addDecorator(withKnobs);

stories.add(
  'Live testing',
  () => {
    const cardTitle = text('Title', 'News');

    const cardType = select(
      'Card type',
      ['general', 'tenant', 'agency'],
      'general',
    );

    const cardDescription = text(
      'Description',
      'Ea qui proident velit cupidatat ut. Ea dolor anim anim id voluptate voluptate incididunt aliqua eu labore.',
    );

    const cardShortcuts = [
      {
        label: text('Shortcut 1 label', 'Shortcut 1'),
        onClick: () => action('button clicked'),
      },
      {
        label: text('Shortcut 2 label', 'Shortcut 2'),
        onClick: () => action('button clicked'),
      },
    ];
    const cardButton = {
      label: text('Button Label', 'Button'),
      onClick: () => action('button clicked'),
    };

    return (
      <Container>
        <Row>
          <Col xs={4} sm={4} md={4} lg={4}>
            <p>Card</p>
            <Card
              image={NewsIcon}
              title={cardTitle}
              description={cardDescription}
              shortcuts={cardShortcuts}
              primaryButton={cardButton}
              type={cardType}
            />
          </Col>
        </Row>
      </Container>
    );
  },
  {
    order: 1,
    info: `
The \`Card\` component is composed by a top image, a title, a description, two shortcut buttons, and an action button.

The component's width is 100%, and it adapts to the width of the element that contains it.

The description text can have a maximum of 150 characters on desktop.  For breakpoints defined for mobile devices (<960 px) the maximum will be 70 characters.

When a user hovers over a card on desktop, the card will highlight and “float” above the page.  Depending on the card type you select, the image background color will be different:

<ul>
    <li>
      General: #C3D6FE
    </li>
    <li>
      Toolkit: #D2EFE5
    </li>
    <li>
      Agency: #FDDBD9
    </li>
</ul>

The default value for the Card type is \`general\`.

Usage instructions:

    import { Card } from '@codelitt/ay-design-library';

    <Card
      image='../images/image.jpg'
      title='Card Title'
      description='Ea qui proident velit cupidatat ut. Ea dolor anim anim id voluptate voluptate incididunt aliqua eu labore.'
      shortcuts=[
        {label: 'Shortcut 1 label', onClick: () => { alert('Hi!')},
        {label: 'Shortcut 2 label', onClick: () => { alert('Hi!')}
      ]
      button={label: 'Button label', onClick: () => { alert('Hi!')}
      type='general' | 'toolkit' | 'leasing'
    />
    `,
  },
);

stories.add(
  'Card',
  () => (
    <Container>
      <Row>
        <Col xs={4} sm={4} md={4} lg={4}>
          <Card
            image={NewsIcon}
            title="News"
            description="This is the card description text"
            shortcuts={[
              {
                label: 'Shortcut 1 label',
                onClick: () => {
                  alert('Hi!');
                },
              },
              {
                label: 'Shortcut 2 label',
                onClick: () => {
                  alert('Hi!');
                },
              },
            ]}
            primaryButton={{
              label: 'Open News',
              onClick: () => {
                alert('Hi!');
              },
            }}
          />
        </Col>
      </Row>
    </Container>
  ),
  {
    order: 2,
  },
);
