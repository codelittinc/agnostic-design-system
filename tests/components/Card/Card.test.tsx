import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Card from '../../../src/components/Card';

describe('Card', () => {
  const getComponent = (onClick?: () => void) =>
    render(
      <Card
        image="../images/image.jpg"
        title="Card Title"
        description="Ea qui proident velit cupidatat ut. Ea dolor anim anim id voluptate voluptate incididunt aliqua eu labore."
        shortcuts={[
          { label: 'Shortcut 1 label', onClick: onClick! },
          { label: 'Shortcut 2 label', onClick: onClick! },
        ]}
        primaryButton={{ label: 'Button label', onClick: onClick! }}
      />,
    );

  const clickOnButton = (buttonLabel: string) => {
    let clickCount = 0;

    const addClick = () => {
      clickCount += 1;
    };

    const { getByText } = getComponent(addClick);
    const node = getByText(buttonLabel);
    fireEvent.click(node);
    return clickCount;
  };

  it('renders without crashing', () => {
    const card = getComponent();
    return expect(card).toMatchSnapshot();
  });

  it('calls onClick prop on Shortcut 1 button click', () => {
    const clickCount = clickOnButton('Shortcut 1 label');
    expect(clickCount).toBe(1);
  });

  it('calls onClick prop on Shortcut 2 button click', () => {
    const clickCount = clickOnButton('Shortcut 2 label');
    expect(clickCount).toBe(1);
  });

  it('calls onClick prop on Button click', () => {
    const clickCount = clickOnButton('Button label');
    expect(clickCount).toBe(1);
  });
});
