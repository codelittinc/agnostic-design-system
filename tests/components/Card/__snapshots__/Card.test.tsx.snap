// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card-container"
      >
        <div>
          <div
            class="image-container general"
          >
            <img
              alt="card-icon "
              class="image"
              src="../images/image.jpg"
            />
          </div>
          <div
            class="title"
          >
            <span
              class="title card-title"
            >
              Card Title
            </span>
          </div>
          <div
            class="description"
          >
            <span
              class="body-content card-description"
            >
              Ea qui proident velit cupidatat ut. Ea dolor anim anim id voluptate voluptate incididunt aliqua eu labore.
            </span>
          </div>
        </div>
        <div>
          <div
            class="shortcut"
          >
            <button
              class="button ghost left-icon small-size"
              type="button"
            >
              Shortcut 1 label
            </button>
          </div>
          <div
            class="shortcut no-border"
          >
            <button
              class="button ghost left-icon small-size"
              type="button"
            >
              Shortcut 2 label
            </button>
          </div>
          <div
            class="button-container"
          >
            <button
              class="button inline regular-size"
              type="button"
            >
              Button label
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card-container"
    >
      <div>
        <div
          class="image-container general"
        >
          <img
            alt="card-icon "
            class="image"
            src="../images/image.jpg"
          />
        </div>
        <div
          class="title"
        >
          <span
            class="title card-title"
          >
            Card Title
          </span>
        </div>
        <div
          class="description"
        >
          <span
            class="body-content card-description"
          >
            Ea qui proident velit cupidatat ut. Ea dolor anim anim id voluptate voluptate incididunt aliqua eu labore.
          </span>
        </div>
      </div>
      <div>
        <div
          class="shortcut"
        >
          <button
            class="button ghost left-icon small-size"
            type="button"
          >
            Shortcut 1 label
          </button>
        </div>
        <div
          class="shortcut no-border"
        >
          <button
            class="button ghost left-icon small-size"
            type="button"
          >
            Shortcut 2 label
          </button>
        </div>
        <div
          class="button-container"
        >
          <button
            class="button inline regular-size"
            type="button"
          >
            Button label
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
